source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc

```diff
@@ -0,0 +1,114 @@
+namespace NetworkFilters {
+namespace DubboProxy {
+namespace {
+
+constexpr uint16_t MagicNumber = 0xdabb;
+constexpr uint8_t MessageTypeMask = 0x80;
+constexpr uint8_t EventMask = 0x20;
+constexpr uint8_t TwoWayMask = 0x40;
+constexpr uint8_t SerializationTypeMask = 0x1f;
+
+} // namespace
+
+// Consistent with the SerializationType
+bool isValidSerializationType(SerializationType type) {
+  switch (type) {
+  case SerializationType::Hessian:
+  case SerializationType::Json:
+    break;
+  default:
+    return false;
+  }
+  return true;
+}
+
+// Consistent with the ResponseStatus
+bool isValidResponseStatus(ResponseStatus status) {
+  switch (status) {
+  case ResponseStatus::Ok:
+  case ResponseStatus::ClientTimeout:
+  case ResponseStatus::ServerTimeout:
+  case ResponseStatus::BadRequest:
+  case ResponseStatus::BadResponse:
+  case ResponseStatus::ServiceNotFound:
+  case ResponseStatus::ServiceError:
+  case ResponseStatus::ClientError:
+  case ResponseStatus::ServerThreadpoolExhaustedError:
+    break;
+  default:
+    return false;
+  }
+  return true;
+}
+
+void RequestMessageImpl::fromBuffer(Buffer::Instance& data) {
+  ASSERT(data.length() >= DubboProtocolImpl::MessageSize);
+  uint8_t flag = data.peekInt<uint8_t>(2);
+  is_two_way_ = (flag & TwoWayMask) == TwoWayMask ? true : false;
+  type_ = static_cast<SerializationType>(flag & SerializationTypeMask);
+  if (!isValidSerializationType(type_)) {
+    throw EnvoyException(
+        fmt::format("invalid dubbo message serialization type {}",
+                    static_cast<std::underlying_type<SerializationType>::type>(type_)));
+  }
+}
+
+void ResponseMessageImpl::fromBuffer(Buffer::Instance& buffer) {
+  ASSERT(buffer.length() >= DubboProtocolImpl::MessageSize);
+  status_ = static_cast<ResponseStatus>(buffer.peekInt<uint8_t>(3));
+  if (!isValidResponseStatus(status_)) {
+    throw EnvoyException(
+        fmt::format("invalid dubbo message response status {}",
+                    static_cast<std::underlying_type<ResponseStatus>::type>(status_)));
+  }
+}
+
+bool DubboProtocolImpl::decode(Buffer::Instance& buffer, Protocol::Context* context) {
+  if (buffer.length() < DubboProtocolImpl::MessageSize) {
+    return false;
+  }
+
+  uint16_t magic_number = buffer.peekBEInt<uint16_t>();
+  if (magic_number != MagicNumber) {
+    throw EnvoyException(fmt::format("invalid dubbo message magic number {}", magic_number));
+  }
+
+  uint8_t flag = buffer.peekInt<uint8_t>(2);
+  MessageType type = static_cast<MessageType>((flag & MessageTypeMask) >> 7);
+  bool is_event = (flag & EventMask) == EventMask ? true : false;
+  int64_t request_id = buffer.peekBEInt<int64_t>(4);
+  int32_t body_size = buffer.peekBEInt<int32_t>(12);
+
+  if (body_size > MaxBodySize || body_size <= 0) {
+    throw EnvoyException(fmt::format("invalid dubbo message size {}", body_size));
+  }
+
+  context->body_size_ = body_size;
+
+  if (type == MessageType::Request) {
+    RequestMessageImplPtr req =
```

Comment:
This can be `RequestMessagePtr` since fromBuffer is virtual, and you don't need to typedef `RequestMessageImplPtr`. same for response below.

Reply:
fromBuffer is MessageImpl virtual interface，not RequestMessage or ResponseMessage

Reply:
Oh ok, then I don't think you need it fromBuffer to be virtual

Reply:
I need to call fromBuffer interface, but RequestMessagePtr don't have this interface, so I used here is RequestMessageImplPtr, can you give me some advice?

Reply:
Leave them in RequestMessageImpl and ResponseMessageImpl, but they don't have to be virtual pure in MessageImpl, you can implement them independenctly.